### DISCUSSION TITLE
Vertical/Portrait UI layout

### DISCUSSION THREAD

**Discussion by crazy-hair** [suchmememanyskill/CYD-Klipper#14] [Ideas]
Just to get this out of the way: I know this is a huge ask and I don't want it to seem like I expect it to happen, but was mainly just curious how feasible it was. 

Now that we have that out of the way, the question is: would it be possible to get a vertical UI which would allow for two things: way better viewing angles (looking at the screen in landscape can hurt your eyes sometimes) and also handheld battery-powered iPhone-4-size-and-style devices that you can operate with one hand and take between your desk and your printer. I plan on designing a case that has pogo pins and have docks both attached to the printer and on my desk so I can just grab it and go, and am trying to determine which orientation would work better. (sidenote, wonder if [in this situation] it would be possible to switch orientations by pulling a GPIO pin to ground or the like, allowing for the screen to be horizontal while docked and vertical while handheld). 

![worlds_shittist_mockupc](https://github.com/suchmememanyskill/CYD-Klipper/assets/9155262/3c229ae2-1c83-47b1-aba1-4a0289130d05)

Thanks for your time and sorry for my ramblings!

**Comment by suchmememanyskill** [comment]
It would essentially mean rewriting all ui code. I have considered it before but it requires essentially if-else-ing all positional data :(

Needless to say, i didn't want to make the codebase more of a mess than it already was and invest time for a feature i'll probably never use, as i made horizontal mounts for my printer

**Comment by suchmememanyskill** [comment]
Ever since the UI rewrite the UI is a lot more responsive, so i thought i'd give this a go (see the experimental-vertical-ui branch, and the smartscreen driver). It sort of works but i'm most certainly not happy with the result

![PXL_20240218_155518020](https://github.com/suchmememanyskill/CYD-Klipper/assets/38142618/ace5b5c8-fdf0-41f0-b551-1e35df4dc110)
![PXL_20240218_155537994](https://github.com/suchmememanyskill/CYD-Klipper/assets/38142618/2a9b1edb-05fa-40b2-ba47-9df587d56dbd)
![PXL_20240218_155557321](https://github.com/suchmememanyskill/CYD-Klipper/assets/38142618/ebee79af-ef7b-44cd-b204-4fa584d1dae4)

**Comment by crazy-hair** [comment]
Wow! Didn't expect any follow up to this, this is great. Is the new driver currently set up in any of the PIO environments? When I build using the regular ENV it throws the "old driver doesn't support vertical" error, and when I compile with the smartdisplay ENV it just displays static. Is there additional configuration I have to go through? Sorry, I'm not too versed in PIO

**Reply by suchmememanyskill** [reply]
You may have to use another configuration from https://github.com/rzeldent/platformio-espressif32-sunton/tree/main depending on what model of screen that you have. The current smartdisplay .json in the boards folder is the v3 (microb + usbc), which may not work with the v2/v1 (usbc, microb, respectively). You do have to transplant some options from the existing smartdisplay json as there's options to configure cyd-klipper related stuff.

Also half the ui is not working as expected, probably not worth using

**Comment by suchmememanyskill** [comment]
@CrazyHairGuy got bored and was unhappy with the first implementation.

Second implementation is on the dev branch. Activate it by replacing the following text in platformio.ini

```
[env:esp32-3248S035C-smartdisplay]
board = esp32-3248S035C-smartdisplay
build_flags = 
	-DLV_CONF_PATH="../../../../src/conf/lv_conf.h"
	-DCYD_SCREEN_VERTICAL=1
	-DTOUCH_SWAP_XY=false
	
[env:esp32-2432S028R-smartdisplay]
board = esp32-2432S028R-smartdisplay
build_flags = 
	-DLV_CONF_PATH="../../../../src/conf/lv_conf.h"
	-DCYD_SCREEN_VERTICAL=1
	-DCYD_SCREEN_DISABLE_TOUCH_CALIBRATION=1
	-DTOUCH_SWAP_XY=false
```

less stuff is broken than first :)

**Reply by crazy-hair** [reply]
Interesting, I still can't get a display output using the smartdisplay driver. The firmware compiles just fine but I don't get an image: 
![PXL_20240302_221227215](https://github.com/suchmememanyskill/CYD-Klipper/assets/9155262/08cc3265-8fd8-439d-b0c6-671194f6913d)
I am using the v3 board, and the dev branch.

**Reply by suchmememanyskill** [reply]
For the fun of it, can you try the [v2](https://github.com/rzeldent/platformio-espressif32-sunton/blob/4aa429fdefeda907a9e3e8d9b0f4a22feda4d152/esp32-2432S028Rv2.json) and [v1](https://github.com/rzeldent/platformio-espressif32-sunton/blob/4aa429fdefeda907a9e3e8d9b0f4a22feda4d152/esp32-2432S028R.json) configs? You'd have to port over the CYD-Klipper specific defines

**Reply by suchmememanyskill** [reply]
Also, maybe worth trying out https://github.com/rzeldent/esp32-smartdisplay-demo to check if it works there. Do note that CYD-Klipper uses v2.0.5 of the smartdisplay driver. 2.0.5 -> 2.0.7 as far as i could tell didn't include many changes for this display

**Reply by crazy-hair** [reply]
When I try the V2 config, I get a compile error: ```xtensa-esp32-elf/bin/ld.exe: .pio/build/esp32-2432S028R-smartdisplay/libd7c/libesp32_smartdisplay.a(lvgl_ili9341_spi.c.o): in function `lvgl_lcd_init':
F:\Tools\esp32\cyd\cyd-klipper-dev\CYD-Klipper\CYD-Klipper/.pio/libdeps/esp32-2432S028R-smartdisplay/esp32_smartdisplay/src/lvgl_ili9341_spi.c:83: undefined reference to `esp_lcd_panel_disp_on_off'collect2.exe: error: ld returned 1 exit status
*** [.pio\build\esp32-2432S028R-smartdisplay\firmware.elf] Error 1```

These are the only differences in the config (it all seems to be related to the different driver chip on the older boards): 
![image](https://github.com/suchmememanyskill/CYD-Klipper/assets/9155262/9468bb59-91fb-4c15-bd98-72b1ca52f3fd)

The demo you linked compiles and runs just fine, though.

**Reply by suchmememanyskill** [reply]
Just making sure, your board works using the v2 driver in the demo app? Or do you mean works with v3 in the demo app?

**Reply by crazy-hair** [reply]
It works with the V3 driver in the demo.

**Reply by suchmememanyskill** [reply]
@CrazyHairGuy i just pushed another commit to dev adding the other 2 versions of the 2.8 inch resistive model. I don't get any compile errors on my side. Give it a shot

**Reply by suchmememanyskill** [reply]
> It works with the V3 driver in the demo.

Oh that's very strange. Also if you downgrade the demo to v2.0.5?

**Reply by crazy-hair** [reply]
Yep, demo compiled at version 2.0.5.

**Comment by suchmememanyskill** [comment]
@crazy-hair do you by chance have a 3.5" capacitive display? I've done a port of the vertical UI to that screen and will probably release that too as it actually looks decent. Currently available in the dev branch

**Reply by crazy-hair** [reply]
I actually have one on order from AliExpress, but it probably won't be here for a few weeks (AliExpress shipping seems to have gotten worse in the past few months?). I'd be glad to test it out once I get it though.

**Comment by crazy-hair** [comment]
3.5" display arrived, flashed the vertical config from the web installer. Gotta say, this is a much better experience than the 2.8r display. Especially with the temp graph and printer list having more room on the display to eliminate some scrolling.
